bite-size-notes/week-4-d-Example of LQR-name.md

```diff
@@ -0,0 +1,350 @@
+Ptf(3,1) = F(2,1);
+Ptf(4,1) = F(2,2);
+nfig = 0;
+options=odeset('RelTol',1e-10);
+[tk,PV]=ode45(@(t,p) dpp(t,p,flag,A,B,Q,R,E,F),[0 (tf-ti)],Ptf,options);
+
+
+% Solving for P in a vector form PV
+%
+%
+% PV is in vector form, each row corresponds to row in time t
+% flip the PV vector 
+PV = flipud(PV);
+% redefine the time vector
+tk = flipud(tk); 
+tk = -tk +(tf)*ones(size(tk));
+%
+%
+% computing the gain matrix K(t) as row vector
+%
+[mP,nP] = size(PV);
+K = zeros(mP,2);
+%
+for i = 1:mP
+Ptm = zeros(2) ; 
+Ptm(1,1) = PV(i,1);
+Ptm(1,2) = PV(i,2);
+Ptm(2,1) = PV(i,3);
+Ptm(2,2) = PV(i,4);
+K(i,:) = inv(R)*B'*Ptm;
+
+end
+
+
+[y2, dyx] = ode45(@(t,x) dyn(t,x,A,B),[0 (tf-ti)],x0,options);
+
+figure 
+plot(y2,dyx,'o','LineWidth',3)
+xlabel('time')
+ylabel('State')
+legend('uncontrol State x1' ,'uncontrol State x2')
+set(gca,'FontSize',20)
+
+% plot the control gain K
+nfig = nfig+1; figure(nfig)
+figure
+plot(tk,K,'r','LineWidth',3)
+xlabel('time')
+ylabel('Kalman gain K')
+set(gca,'FontSize',20)
+
+%return
+% simulate the sytem response using ode45
+options=odeset('RelTol',1e-10);
+[t,x]=ode45(@(t,x) xdiff(t,x,flag,A,B,tk,K),[0 (tf-ti)],x0,options);
+% plot the state x
+nfig = nfig+1; figure(nfig)
+figure
+plot(t,x,'b','LineWidth',3)
+xlabel('time')
+ylabel('State')
+legend('Controled State x1' ,'Controled State x2')
+set(gca,'FontSize',20)
+
+%return
+% find and plot the input u
+
+[m,n]=size(x);     % Length of time, x is m
+[mR,nR] = size(R); % number of inputs = mR
+Kt = interp1(tk,K,t); % interpolate the K matrix
+u = zeros(m,mR);    % initialize the input
+for jj=1:1:m
+    u(jj) = -Kt(jj,:)*(x(jj,:))';
+end
+
+figure
+plot(t,u,'r','LineWidth',3)
+xlabel('time')
+ylabel('Input u')
+set(gca,'FontSize',20)
+
+
+
+
+
+
+
+
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%%% Function to find the solution to Riccati Eq
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+function dpv = dpp(t,p,flag,A,B,Q,R,E,F)
+%
+%    Called by PSolve to compute the Riccati matrix P.
+%
+%    [t,GP] = ode45('gramdiff',[ti tf],zeros(NT,1),[],A,B,Q,R,E).
+%
+%    See also PSOLVE.
+%
+%
+% Finding the P matrix PM
+% 
+PM = zeros(2);
+PM(1,1) = p(1);
+PM(1,2) = p(2);
+PM(2,1) = p(3);
+PM(2,2) = p(4);
+%
+%
+% computing P matrix derivative 
+% Backward in time
+dp = A'*PM + PM*A -PM*E*PM +Q;
+%
+%
+% computing P vector derivative
+dpv = zeros(4,1);
+
+dpv(1,1) = dp(1,1) ;
+dpv(2,1) = dp(1,2) ;
+dpv(3,1) = dp(2,1) ;
+dpv(4,1) = dp(2,2) ;
+end
+
+
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%%%% Function to find the solution to system with optimal control
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+function xdiff = xdiff(t,x,flag,A,B,tK,K)
+%
+% solving xdot = AX + Bu
+
+Kt = interp1(tK,K,t);
+ut = -Kt*x;
+xdiff = A*x +B*ut;
+end
+
+function dx = dyn(t,x,A,B)
+
+dx = A*x; 
+end
+```
+
+# Results: 
+First we can we plot the states without control, and find the system blows up, since the eigenvalue of the A matrix is on the RHP.
```

Comment:
Because eigenvalue of A is on the RHP, so the system is unstable and state goes to infinity

Reply:
yes, my point was just that eigenvalues is not super intuitive in time domain analysis, so it may be helpful to write out explicitly what the values of those eigenvalues are because you have a good numerical example example!